<html>
<head>
<title>Ensembl: Data Tools - Assembly converter, ID History converter, Variant Effect Predictor</title>
</head>
<body>

<p>We provide a number of ready-made tools for processing your data. At the moment, small datasets can be uploaded to our servers and processed online; for larger datasets, we provide an API script that can be downloaded (you will also need to <a href="http://www.ensembl.org/info/docs/api/api_installation.html">install our Perl API</a> to use these).</p>
<p>In the near future we aim to offer an intermediate service, whereby medium-to-large data sets can be submitted to a queue, similar to BLAST.</p>
<p>Currently available:</p>

<table class="ss" style="width:100%">
<tr>
  <th style="width:20%">Tool</th>
  <th style="width:40%">Description</th>
  <th style="width:20%">&nbsp;</th>
  <th style="width:20%">&nbsp;</th>
  <th style="width:20%">&nbsp;</th>
</tr>

<tr>
  <td>Assembly converter</td>
  <td>Map your data to the current assembly. Accepted file formats: 
  <a href="/info/website/upload/gff.html">GFF</a>, 
  <a href="/info/website/upload/gff.html">GTF</a>, 
  <a href="/info/website/upload/bed.html">BED</a>, 
  <a href="/info/website/upload/psl.html">PSL</a>
  <br />N.B. Export is currently in GFF only</td>
  <td><a href="http://grch37.ensembl.org/Homo_sapiens/Tools/AssemblyConverter?db=core" class="modal_link-no">Ensembl version</a></td>
  <td><a href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/SelectFeatures" class="modal_link-no">Online version</a></td>
  <td><a href="ftp://ftp.ensembl.org/pub/misc-scripts/Assembly_mapper_1.0/">API script</a></td>
</tr>

<tr class="bg2">
  <td>ID History converter</td>
  <td>Convert a set of Ensembl IDs from a previous release into their current equivalents.</td>
  <td><a href="http://grch37.ensembl.org/Homo_sapiens/Tools/IDMapper?db=core" class="modal_link-no">Ensembl version</a></td>
  <td><a href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/UploadStableIDs" class="modal_link-no">Online version</a><br />(max 30 ids)</td>
  <td><a href="ftp://ftp.ensembl.org/pub/misc-scripts/ID_mapper_1.0/">API script</a></td>
</tr>

<tr>
  <td>Variant Effect Predictor</td>
  <td>(Formerly SNP Effect Predictor). Upload a set of SNPs in our <a href="/info/website/upload/var.html">standard format</a> and export a file containing consequence types. Uploaded tracks can also be viewed on Location pages.</td>
  <td><a href="http://grch37.ensembl.org/Homo_sapiens/Tools/VEP?db=core" class="modal_link-no">Ensembl version</a></td>
  <td><a href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/UploadVariations" class="modal_link-no">Online version</a><br />(max 750 SNPs)</td>
  <td><a href="ftp://ftp.ensembl.org/pub/misc-scripts/Variant_effect_predictor/">API script</a></td>
</tr>
<tr class="bg2">
  <td>Data Slicer</td>
  <td>Get a subset of data from a BAM or VCF file.</td>
  <td><a href="http://grch37.ensembl.org/Homo_sapiens/Tools/DataSlicer?db=core" class="modal_link-no">Ensembl version</a></td>
  <td><a href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/SelectSlice" class="modal_link-no">Online version</a><br />(max 10K region)</td>
  <td><a href="http://www.1000genomes.org/faq/how-do-i-get-sub-section-vcf-file">Command line options</a></td>
</tr>
<tr class="bg2">
  <td>Variation Pattern Finder</td>
  <td>Identify variation patterns in a chromosomal region of interest for different individuals. Only variations with functional significance such non-synonymous coding, splice site will be reported by the tool. Click <a href="http://www.1000genomes.org/variation-pattern-finder" target=_blank>here</a> for more extensive documentation. </td>
  <td><a href="http://grch37.ensembl.org/Homo_sapiens/Tools/VariationPattern?db=core" class="modal_link-no">Ensembl version</a></td>
  <td><a href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/VariationsMapVCF" class="modal_link-no">Online version</a><br /></td>
  <td><a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/browser/variation_pattern_finder/version_1.0">API script</a></td>
</tr>

<tr class="bg2">
  <td>VCF to PED converter</td>
  <td>The VCF to PED converter allows users to parse a vcf file to create a linkage pedigree file (ped) and a marker information file, which together may be loaded into ld visualization tools like Haploview. Click <a href="http://www.1000genomes.org/vcf-ped-converter" target=_blank>here</a> for more extensive documentation.</td>
  <td><a href="http://grch37.ensembl.org/Homo_sapiens/Tools/VcftoPed?db=core" class="modal_link-no">Ensembl version</a></td>
  <td><a href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/Haploview" class="modal_link-no">Online version</a><br /></td>
  <td><a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/browser/vcf_to_ped_converter/version_1.1/">API script</a></td>
</tr>

<tr class="bg2">
  <td>Forge analyser</td>
  <td>The Forge tool takes a list of variants and analizes their enrichment in functional regions from the ENCODE or Roadmap Epigenome project on a tissue specific basis. The analysis requires a miniumum of 20 variants which have to be present in phase 1 of 1000 Genomes project. Click <a href="http://www.1000genomes.org/forge-analysis-11" target="_blank">here</a> for full tool documentation.</td>
  <td>-</td>
  <td><a href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/Forge" class="modal_link-no">Online version</a><br /></td>
  <td><a href="https://github.com/iandunham/Forge.git">API script</a></td>
</tr>

<tr class="bg2">
  <td>Allele frequency calculator</td>
  <td>This tool takes a VCF file, a matching sample panel file, a chromosomal region, a population name and calculates population-wide allele frequency for sites within the chromosomal region defined.</td>
  <td><a href="http://grch37.ensembl.org/Homo_sapiens/Tools/AlleleFrequency?db=core" class="modal_link-no">Ensembl version</a></td>
  <td><a href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/Allele" class="modal_link-no">Online version</a><br /></td>
  <td><a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/browser/allele_frequency_calculator/">API script</a></td>
</tr>
</table>

</body>
</html>
